File: app/src/main/java/knf/kuma/animeinfo/AnimeRelatedAdapter.java
package knf.kuma.animeinfo;

import android.content.Context;
import android.support.v4.app.Fragment;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import knf.kuma.R;
import knf.kuma.commons.PatternUtil;
import knf.kuma.commons.PicassoSingle;
import knf.kuma.database.CacheDB;
import knf.kuma.database.dao.AnimeDAO;
import knf.kuma.pojos.AnimeObject;

public class AnimeRelatedAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder>{
    private Context context;
    private Fragment fragment;
    private List<AnimeObject.WebInfo.AnimeRelated> list=new ArrayList<>();
    private AnimeDAO dao= CacheDB.INSTANCE.animeDAO();

    public AnimeRelatedAdapter(Fragment fragment, List<AnimeObject.WebInfo.AnimeRelated> list) {
        this.context=fragment.getContext();
        this.fragment=fragment;
        this.list = list;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        switch (viewType) {
            case 0:
                return new RelatedHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_related, parent, false));
                default:
            case 1:
                return new RelatedNoImgHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_related_noimg,parent,false));
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder h, int position) {
        final AnimeObject.WebInfo.AnimeRelated related=list.get(position);
        if (h.getItemViewType()==0){
            final RelatedHolder holder=(RelatedHolder)h;
            final AnimeObject object=dao.getByLink("%"+related.link);
            PicassoSingle.get(context).load(object.img).into(holder.imageView);
            holder.textView.setText(related.name);
            holder.relation.setText(related.relation);
            holder.cardView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    ActivityAnime.open(fragment,object,holder.imageView);
                }
            });
        }else {
            final RelatedNoImgHolder holder=(RelatedNoImgHolder) h;
            holder.textView.setText(related.name);
            holder.relation.setText(related.relation);
            holder.cardView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    ActivityAnime.open(fragment,related);
                }
            });
        }
    }

    @Override
    public int getItemViewType(int position) {
        return (dao.existSid(PatternUtil.getLinkNumber(list.get(position).link))?0:1);
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    class RelatedHolder extends RecyclerView.ViewHolder{
        @BindView(R.id.card)
        CardView cardView;
        @BindView(R.id.img)
        ImageView imageView;
        @BindView(R.id.title)
        TextView textView;
        @BindView(R.id.relation)
        TextView relation;

        public RelatedHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this,itemView);
        }
    }

    class RelatedNoImgHolder extends RecyclerView.ViewHolder{
        @BindView(R.id.card)
        CardView cardView;
        @BindView(R.id.title)
        TextView textView;
        @BindView(R.id.relation)
        TextView relation;

        public RelatedNoImgHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this,itemView);
        }
    }
}

File: app/src/main/java/knf/kuma/animeinfo/AnimeRelatedAdapter.kt
package knf.kuma.animeinfo

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import knf.kuma.R
import knf.kuma.commons.PatternUtil
import knf.kuma.commons.PicassoSingle
import knf.kuma.databinding.ItemRelatedBinding
import knf.kuma.pojos.AnimeObject

internal class AnimeRelatedAdapter(private val fragment: Fragment, private val list: MutableList<AnimeObject.WebInfo.AnimeRelated>) : RecyclerView.Adapter<AnimeRelatedAdapter.RelatedHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RelatedHolder {
        return RelatedHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_related, parent, false))
    }

    override fun onBindViewHolder(holder: RelatedHolder, position: Int) {
        val related = list[position]
        holder.textView.text = related.name
        holder.relation.text = related.relation
        if (related.aid != "null") {
            holder.imageView.visibility = View.VISIBLE
            PicassoSingle.get().load(PatternUtil.getCover(related.aid)).into(holder.imageView)
            holder.cardView.setOnClickListener { ActivityAnime.open(fragment, related, holder.imageView) }
        } else {
            holder.imageView.visibility = View.GONE
            holder.cardView.setOnClickListener { ActivityAnime.open(fragment, related) }
        }
    }

    override fun getItemCount(): Int {
        return list.size
    }

    internal class RelatedHolder(itemView: View, binding: ItemRelatedBinding = ItemRelatedBinding.bind(itemView)) : RecyclerView.ViewHolder(itemView) {
        val cardView: LinearLayout = binding.card
        val imageView: ImageView = binding.img
        val textView: TextView = binding.title
        val relation: TextView = binding.relation
    }
}
